fms/modules/tp.py

```diff
@@ -31,54 +25,205 @@ def setup_tp(self, rank: int, world_size: int) -> None:
+    ):
+        """
+        This function copies the correct shard of the weights for a rowwise-TP'd module
+        according to the rank of the process and the world_size.
 
-    @staticmethod
-    @abstractmethod
-    def import_module(module, group: ProcessGroup):
-        pass
+        Args
+        ====
+        param: torch.nn.Parameter
+            Parameter that has had TP applied
+        tensor_value: torch.Tensor
+            tensor that needs sharding
+        is_bias: bool
+            if True is bias, else is weight
+        max_partition_sizes: List[int]
+            for each number in the list, if world_size is smaller than or equal to that number, the tensor will get
+            partitioned in worldsize parts, else if world size is larger than the number then you will get world size parts
+            replicated in worldsize / number
+
+            world_size = 4, max_partition_sizes = [8], tensor = [0 1 2 3 4 5 6 7]
+            [0 1] [2 3] [4 5] [6 7]
+
+            world_size = 8, max_partition_sizes = [4], tensor = [0 1 2 3 4 5 6 7]
+            [0 1] [0 1] [2 3] [2 3] [4 5] [4 5] [6 7] [6 7]
+
+            If there are multiple numbers in the max_partition_sizes list, then the param gets filled with non-contiguous
+            slices of the tensor_value. This is useful for fused weight cases (qkv, mlp, moe, etc.)
+
+            world_size = 4, max_partition_sizes = [4, 4], tensor = [0 1 2 3 4 5 6 7]
+            [0 4] [1 5] [2 6] [3 7]
+
+            world_size = 4, max_partition_sizes = [4, 1], tensor = [0 1 2 3 4 5 6 7 8 9]
+            [0 1 8 9] [2 3 8 9] [4 5 8 9] [6 7 8 9]
+        """
+        # Divide the weight matrix along the second dimension.
+        output_size_per_partition = param.shape[1] // (
+            sum(max_partition_sizes) // min(max_partition_sizes)
+        )
+        if not is_bias:
+            tp_slices = self.__get_tp_slices(
+                tensor_value.shape[1], output_size_per_partition, max_partition_sizes
+            )
+            tensors_to_cat = [tensor_value[:, tp_slice] for tp_slice in tp_slices]
+            tensor = torch.cat(tensors_to_cat, dim=1)
+            param.copy_(tensor, non_blocking=True)
+        else:
+            if self.rank == 0:
+                param.copy_(tensor_value, non_blocking=True)
+            else:
+                param.zero_()
+
+    def copy_colwise(
+        self,
+        param: Union[torch.nn.Parameter, torch.Tensor],
+        tensor_value: torch.Tensor,
+        is_bias: bool,
+        max_partition_sizes: List[int],
+    ):
+        """
+        This function copies the correct shard of the weights for a colwise-TP'd module
+        according to the rank of the process and the world_size.
+
+        Args
+        ====
+        param: torch.nn.Parameter
+            Parameter that has had TP applied
+        tensor_value: torch.Tensor
+            tensor that needs sharding
+        is_bias: bool
+            if True is bias, else is weight
```

Comment:
it might help to have a clearer explanation of the treatment/sharding of bias vs non-bias values during the copy. It might even be better to have the paramter name signify the difference in the operation vs the difference in how it's being used (maybe there could be other modules with paramters that are treated like bias, but may not literally be 'bias' terms?)

Reply:
yup, reworded to a better name with better comment too